### ISSUE TITLE
output validation

### ISSUE THREAD

**Issue by fcollman** [AllenInstitute/argschema#38]
We have setup the default arguments to include --output_json, but we haven't actually done anything with it by default.  

I propose that we make an extension of ArgSchemaParser, ArgSchemaOutputParser which standardizes how this would be done, but makes it's use optional for those who want it.  Maybe even the output_json should do with the schema_type, so that output_json isn't a default part of the ArgSchema unless you are using ArgSchemaOutputParser... but i'm a bit confused on that point.

```Python
class ArgSchemaOutputParser(ArgSchemaParser):

    def __init__(self,output_schema_type = None, *args, **kwargs):
        self.output_schema_type = output_schema_type
        super(ArgSchemaOutputParser,self).__init__(*args,**kwargs)
    
    def output(self,d):
        """outputs a dictionary to the output_json file path after
        validating it through the output_schema_type

        Parameters
        ----------
        d:dict
            output dictionary to output to self.mod['output_json'] location
        
        Raises:
        mm.ValidationError
        """

        schema = self.output_schema_type()
        (output_json,errors)=schema.dump(d)
        if len(errors)>0:
            raise mm.ValidationError(json.dumps(errors))
        with open(self.args['output_json'],'w') as fp:
            json.dump(output_json,fp)

```

**Comment by dyf** [comment]
I vote for not doing this in a subclass.  It will always optional since the user doesn't actually have to call `parser.output()`.

**Comment by fcollman** [comment]
and i suppose that output_schema_type can be default=None and it won't matter unless they try to use output without setting it, and we can just skip validation if output_schema_type is None, maybe warn them.

**Comment by JohnGalbraith** [comment]
So how would this work if I wanted to drop an output not in json?  Like a big binary output in HDF5 or an image?  That would be valid output too, right?

**Comment by dyf** [comment]
The module can write whatever files it wants.   I use the output JSON for parameter interchange / communicating in a standard way with databases / etc.  Might as well schematize and validate it since the infrastructure is there.

**Comment by fcollman** [comment]
```Python
    def output(self,d):
        """method for outputing dictionary to the output_json file path after
        validating it through the output_schema_type

        Parameters
        ----------
        d:dict
            output dictionary to output to self.mod['output_json'] location
        
        Raises:
        mm.ValidationError
        """
        if self.output_schema_type is not None:
            schema = self.output_schema_type()
            (output_json,errors)=schema.dump(d)
            if len(errors)>0:
                raise mm.ValidationError(json.dumps(errors))
        else:
            self.logger.warning("output_schema_type is not defined,\
                                 the output won't be validated")
            output_json = d

        with open(self.args['output_json'],'w') as fp:
            json.dump(output_json,fp)
```

**Comment by fcollman** [comment]
@JohnGalbraith  yeah i think the idea would be you would write the hdf5 file yourself.  And then your output_schema would specify that one of the outputs is a path the an hdf5 file, which the output_schema would then validate is a valid hdf5 file if you had implemented that field type.

**Comment by fcollman** [comment]
closed via #39
